RFC: Use Percy/Cypress for integration-style visual regression testing (#5287)
guardian/dotcom-rendering: dotcom-rendering/docs/architecture/030-use-percy-for-e2e-visual-regression-testing.md

# Use Percy/Cypress for integration-style visual regression testing

> ## Status: Proposed

Reviewer: Once it’s merged in… it’s accepted, isn’t it? I’m not sure what the purpose of “status” is actually…

```suggestion
## Status: Accepted
```

Author: that's a good question! Initially I'd been thinking that it would be merged as a proposal, but given the feedback so far it seems like it makes sense to just add it as an accepted proposal. I'll add a `Decision` section today and then should be able to merge it unless anything else comes up

  which depends on a salt value set in `.env`.
- Cypress makes it easier to test a larger slice of the rendering flow.
  - Example use-case: Changes to the `enhance` methods like
  [`enhanceCollections`](https://github.com/guardian/dotcom-rendering/blob/1b37daa385aa348d3ac666d81ba0f666f56bf577/dotcom-rendering/src/model/enhanceCollections.ts#L4)
  can have a significant visual impact, because they control which
  props are forwarded to the React components which render collections.
  Because many of the props are optional, none of the existing tests
  or type checks will catch these changes, and contributors may not
  expect their edit to have visual impacts because `enhance` is one
  step removed from the actual rendering functions. But currently our
  Storybook tests [will not catch](https://github.com/guardian/dotcom-rendering/pull/5119#issuecomment-1147538238)
  this kind of change.
- Because Cypress can easily do integration tests, this should in 
  theory make it easier to visually test states which result from
  user interaction, without needing to write separate stories for each
  state.

> - Because Percy is part of Browserstack, it should make cross-browser   testing easy.

Reviewer 2: ```suggestion
- Because Percy is part of Browserstack, it should enable us to make use of cross-browser testing
```

> - Commercial have also 

Reviewer 2: ```suggestion
- The commercial-dev team have also 
```

[expressed interest](https://github.com/guardian/dotcom-rendering/pull/5256). They want to snapshot the page to verify ad placement but need the page loaded by a real browser to be sure they are capturing a realistic experience
### Cons

- Even though the code footprint of Percy when it's being added to
  Cypress is fairly light, this would nevertheless introduce
  dependency on a new service and set of packages.
- Percy seems to be a relatively new service, and the documentation
  is not currently as detailed as one might hope (although it does
  seem to have [solid backing](https://www.browserstack.com/blog/browserstack-has-acquired-percy/),
  at the time of writing.

> - Cost?

Reviewer 2: ```suggestion
- Cost impact is unknown
```

  Because many of the props are optional, none of the existing tests
  or type checks will catch these changes, and contributors may not
  expect their edit to have visual impacts because `enhance` is one
  step removed from the actual rendering functions. But currently our
  Storybook tests [will not catch](https://github.com/guardian/dotcom-rendering/pull/5119#issuecomment-1147538238)
  this kind of change.
- Because Cypress can easily do integration tests, this should in 
  theory make it easier to visually test states which result from
  user interaction, without needing to write separate stories for each
  state.
- Because Percy is part of Browserstack, it should make cross-browser   testing easy.

### Cons

- Even though the code footprint of Percy when it's being added to
  Cypress is fairly light, this would nevertheless introduce
  dependency on a new service and set of packages.
- Percy seems to be a relatively new service, and the documentation
  is not currently as detailed as one might hope (although it does
  seem to have [solid backing](https://www.browserstack.com/blog/browserstack-has-acquired-percy/),
  at the time of writing.
- Cost?

> 

Reviewer 3: ```suggestion
- Time to execute. This is an unknown quantity at this stage but if the execution time for Percy is too long it could delay deployments
```
